### PR TITLE
Add fromW16CString, rename fromByteStr

### PR DESCRIPTION
**Pull request by harendra-kumar**
Deprecate fromByteStr and rename it to fromCString

### CODE REVIEW

File: core/src/Streamly/Internal/Data/MutArray/Type.hs [hunk 1]
```diff
@@ -1003,7 +1006,7 @@ rightSize arr@MutArray{..} = do
 
 -- XXX We can possibly use a smallMutableByteArray to hold the start, end,
 -- bound pointers.  Using fully mutable handle will ensure that we do not have
--- multiple references to the same array of different lengths lying around and
+-- multiple references to the same array of different Dlengths lying around and
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/MutArray/Type.hs:1009]:
```suggestion
-- multiple references to the same array of different lengths lying around and
```

File: core/src/Streamly/Internal/Data/Stream/Generate.hs [hunk 2]
```diff
@@ -1178,16 +1182,33 @@ fromPtrN n = take n . fromPtr
...
+-- | Read Word16 from an immutable 'Addr#' until a 0 Word16 is encountered, the
+-- 0 Word16 is not included in the stream.
+--
+-- >>> :set -XMagicHash
+-- >>> fromW16CString# addr = Stream.takeWhile (/= 0) $ Stream.fromPtr $ Ptr addr
+--
+-- /Unsafe:/ The caller is responsible for safe addressing.
+--
+{-# INLINE fromW16CString# #-}
+fromW16CString# :: Monad m => Addr# -> Stream m Word8
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:1213]: Shouldn't this be `Stream m Word16`?
According to the doc, read until a `until a 0 Word16 is encountered`.
Here, we stop when a 0 `Word8` is encountered.

File: core/src/Streamly/Internal/Data/Stream/Generate.hs [hunk 3]
```diff
@@ -1178,16 +1182,33 @@ fromPtrN n = take n . fromPtr
...
 {-# INLINE fromByteStr# #-}
 fromByteStr# :: Monad m => Addr# -> Stream m Word8
-fromByteStr# addr = takeWhile (/= 0) $ fromPtr $ Ptr addr
+fromByteStr# = fromCString#
+
+-- | Read Word16 from an immutable 'Addr#' until a 0 Word16 is encountered, the
+-- 0 Word16 is not included in the stream.
+--
+-- >>> :set -XMagicHash
+-- >>> fromW16CString# addr = Stream.takeWhile (/= 0) $ Stream.fromPtr $ Ptr addr
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:1208]: -- >>> fromW16CString# addr = Stream.takeWhile (/= 0) $ Stream.fromPtr (Ptr addr :: Ptr Word16)

Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:1208]:
```suggestion
-- >>> fromW16CString# addr = Stream.takeWhile (/= 0) $ Stream.fromPtr (Ptr addr :: Ptr Word16)
```

File: core/src/Streamly/Internal/Data/Stream/Generate.hs (hunk 2 above)
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/Stream/Generate.hs:1213]: Current implementation is exactly the same as `fromCString#`

File: core/src/Streamly/Internal/Data/MutArray/Type.hs [hunk 4]
```diff
@@ -2459,6 +2462,30 @@ fromByteStr# addr = do
...
+-- | Copy a C string consisting of 16-bit wide chars and terminated by a 16-bit
+-- null char, into a Word16 array. The null character is not copied.
+--
+-- Useful for copying UTF16 strings on Windows.
+--
+{-# INLINABLE fromW16CString# #-}
+fromW16CString# :: MonadIO m => Addr# -> m (MutArray Word16)
+fromW16CString# addr = do
+    w16len <- liftIO $ D.fold FL.length $ D.fromW16CString# addr
+    let bytes = w16len * 2
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/MutArray/Type.hs:2479]: See comment near `fromW16CString#`. Accordingly, this will need to change.

**Review by adithyaov — CHANGES_REQUESTED** [review]
Some comments.

File: core/src/Streamly/Internal/Data/Stream/Generate.hs (hunk 2 above)
Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/Stream/Generate.hs:1213]: This is fixed in the original dir PR, need to pull it here.

**Review by adithyaov — APPROVED** [review]
Looks good.

**Pull request merged**